### PR TITLE
[DA] Refactor `eager` and `on_cron`

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

The goal here is to make the implementations of these conditions simpler, such that it is less daunting to copy/paste/modify. Building up an "intermediate layer" of AutomationConditions which do common / important tasks (e.g. being true if a parent is updated) is an important piece of the overall puzzle. These composite conditions roughly map to "AutoMaterializeRules" in the old world. The advantage is just that these are composed of simpler parts, which can be inspected if desired, and modified at will.

The most controversial name here is probably "since_last_handled". Very open to other suggestions there.

### CODE REVIEW

**Comment by OwenKephart** [comment]
* **#23529** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23529?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by schrockn** [comment]
The PRs a bit noisy. Could you summarize the consolidations that are happening here?

**Comment by OwenKephart** [comment]
@schrockn Yeah I think the diff algorithm kinda choked here. Basically added the following methods:

- since_last_handled
- newly_missing
- any_deps_updated
- any_deps_missing
- any_deps_in_progress
- all_deps_updated_since_cron

and then rewrote `eager` and `on_cron` to use those methods. The goal is for someone reading the definition of `eager` to be able to easily understand what's going on and be able to modify it without having to fully understand the base level primitives

**Review by schrockn — COMMENTED** [review]
As an internal refactor of eager etc it makes sense. However I'm concerned about the surface area here. It kind of leads us to a path where we have to name every arbitrary combination of conditions and support them forever.

**Comment by OwenKephart** [comment]
The basic issue I'm trying to solve here is making it easy for a user to rebuild something similar to `AutomationCondition.eager()`, but with some small modifications. So being able to just do `AutomationCondition.any_deps_missing().ignore(...)` vs:

```python
AutomationCondition.any_deps_match(
    AutomationCondition.missing() & ~AutomationCondition.will_be_requested()
).ignore(...)
```

within their new condition is pretty valuable in terms of ergonomics and simplicity. In general, I want the typical user experience of creating an AutomationCondition "from scratch" to remain fairly high level (i.e. they don't need to think about the fact that we're doing fancy stuff to treat partitions that will be requested as "not missing").

The simple fact that we're using each of these combination of conditions in at least one "endorsed" top-level condition makes me less nervous about supporting these in the long run.

I think:

- any_deps_updated
- any_deps_missing
- any_deps_in_progress
- all_deps_updated_since_cron

are pretty defensible as standalone conceptual entities (all of these have analogous `AutoMaterializeRules`) but the other ones are mostly for convenience and definitely less necessary.

**Review by schrockn — APPROVED** [review]
